### PR TITLE
Read small integers as float32, not float64

### LINKED ISSUE #1842: DataArray read from netcdf with unexpected type
#### Code Sample, a copy-pastable example if possible

```python
import numpy as np
import h5netcdf

filename = "mask_and_scale_float32.nc"

with h5netcdf.File(filename, 'w') as f:
    f.dimensions = {'x': 5}
    v = f.create_variable('hello', ('x',), dtype=np.uint16)
    v[:] = np.ones(5, dtype=np.uint16)
    v[0] = np.uint16(65535)
    v.attrs['_FillValue'] = np.uint16(65535)
    v.attrs['scale_factor'] = np.float32(2)
    v.attrs['add_offset'] = np.float32(0.5)


import xarray as xr

v = xr.open_dataset(filename, mask_and_scale=True)['hello']
print(v.dtype)
```

#### Problem description

The `scale_factor` and `add_offset` being float32, I would expect the result from loading to be a float32 array. However, we get a float64 array instead. A float32 array for a very large dataset is better for faster computations.

#### Expected Output

float32

#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.5.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-693.11.6.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: None.None

xarray: 0.10.0
pandas: 0.21.0
numpy: 1.13.3
scipy: 0.18.1
netCDF4: 1.1.8
h5netcdf: 0.4.2
Nio: None
bottleneck: None
cyordereddict: None
dask: 0.16.0+37.g1fef002
matplotlib: 2.1.0
cartopy: None
seaborn: None
setuptools: 38.2.4
pip: 9.0.1
conda: None
pytest: 3.1.3
IPython: 5.5.0
sphinx: 1.3.6

</details>

### PR DESCRIPTION
**Pull request by Zac-HD**
- [x] Closes #1842
 - [x] Tests added
 - [x] Tests passed
 - [x] Passes ``flake8 xarray`` (now part of tests)
 - [x] Fully documented, including `whats-new.rst` for all changes

Most satellites produce images with color depth in the range of eight to sixteen bits, which are therefore often stored as unsigned integers (with the quality mask in another variable).  If you're lucky, they also have a `scale_factor` attribute and Xarray can automatically convert the integers to floats representing albedo.

This is fantastically convenient, and avoids all the bit-depth bugs from misremembered specifications.  However, loading data as float64 when float32 is sufficient *doubles* memory usage in IO (even on multi-TB datasets...).  While immediately downcasting helps, it's no substitute for doing the right thing first.

So this patch does some conservative checks, and if we can be sure float32 is safe we use that instead.

### CODE REVIEW

File: doc/whats-new.rst [hunk 1]
```diff
@@ -50,6 +50,9 @@ Enhancements
 - :py:func:`~plot.line()` learned to draw multiple lines if provided with a
   2D variable.
   By `Deepak Cherian <https://github.com/dcherian>`_.
+- Reduce memory usage when decoding a variable with a scale_factor, by
+  converting 8-bit and 16-bit integers to float32 instead of float64.
+  By `Zac Hatfield-Dodds <https://github.com/Zac-HD>`_.
```
Comment (jhamman) [inline doc/whats-new.rst]: Add this PR number as a reference.

File: xarray/coding/variables.py [hunk 2]
```diff
@@ -212,11 +212,24 @@ class CFScaleOffsetCoder(VariableCoder):
...
 
+    @staticmethod
+    def _choose_float_dtype(data, attributes):
+        # We default to en/decoding as float64, but for small integer types
+        # and no offset it's usually safe to use float32, which saves a lot
+        # of memory for eg. TB-scale satellite imagery collections.
+        if data.dtype.itemsize <= 2 and \
+                np.issubdtype(data.dtype, np.integer) and \
+                'add_offset' not in attributes and \
+                2 ** -23 < float(attributes.get('scale_factor', 1)) < 2 ** 8:
```
Comment (jhamman) [inline xarray/coding/variables.py]: This if statement is fairly complicated and its not really clear what is happening, can you

1) add some more comments
2) use parentheses instead of line continuation breaks
3) consider just passing in the scale factor instead of attributes

**Comment by Zac-HD** [comment]
Thanks - I was actually writing up an issue and decided it would be easier to demonstrate the proposed fix in a PR, but I'll open an issue first next time.

The checkbox about flake8 could be removed from the issue template now - since #1824 we run flake8 on everything in CI so if tests pass flake8 is passing too.

Re: tests: what do you (and @shoyer) think about using [Hypothesis](https://hypothesis.readthedocs.io) for some property-based tests of variable coding?  "encoding then decoding is a no-op" is a classic property 😄  Upside, more powerful and better at finding edge cases; downside slower simply because it checks more cases (a configurable number).

**Comment by shoyer** [comment]
As mentioned in https://github.com/pydata/xarray/issues/1842, maybe we should also make a point not to upcast float32 input?

One possible concern with changing precision from float64 -> float32 is that some reduce operations like mean could become due to lower precision. So it's a good think @fujiisoup wrote https://github.com/pydata/xarray/pull/1841 so we can specify `dtype` in reductions :).

With regards to Hypothesis: I haven't used it, but it does seem very intriguing. I'm sure that it could turn up quite a few bugs in xarray. Would it make sense to add it as an optional dependency to the test suite? Even if we use Hypothesis to cover more edge cases, I think we will still want normal test coverage for most behavior.

**Comment by Zac-HD** [comment]
Added tests; float32 not upcast, float16 (["intended for storage of many floating-point values where higher precision is not needed, not for performing arithmetic"](https://en.wikipedia.org/wiki/Half-precision_floating-point_format)) *is* upcast but only to float32.

I'll open a new issue to add a basic suite of property-based tests to Xarray 😄

File: doc/whats-new.rst [hunk 3]
```diff
@@ -51,6 +51,10 @@ Enhancements
 - :py:func:`~plot.line()` learned to draw multiple lines if provided with a
   2D variable.
   By `Deepak Cherian <https://github.com/dcherian>`_.
+- Reduce memory usage when decoding a variable with a scale_factor, by
```
Comment (shoyer) [inline doc/whats-new.rst]: It looks like this will also effect encoding as well, e.g., writing float32 rather than float64 by default for scale offset encoded float32 data? Let's note that.

File: xarray/tests/test_coding.py [hunk 4]
```diff
@@ -36,3 +38,14 @@ def test_coder_roundtrip():
...
+
+@pytest.mark.parametrize('dtype', 'u1 u2 i1 i2 f2 f4'.split())
+def test_scaling_converts_to_float32(dtype):
+    original = xr.Variable(('x',), np.arange(10, dtype=dtype),
+                           encoding=dict(scale_factor=10))
+    coder = variables.CFScaleOffsetCoder()
+    encoded = coder.encode(original)
+    assert encoded.dtype == np.float32
+    roundtripped = coder.decode(encoded)
+    assert_identical(original, roundtripped)
```
Comment (shoyer) [inline xarray/tests/test_coding.py]: Let's also add an explicit check that the decoded values are float32 (I don't think assert_identical does that).

File: xarray/coding/variables.py [hunk 5]
```diff
@@ -212,11 +212,31 @@ class CFScaleOffsetCoder(VariableCoder):
         decode_values = encoded_values * scale_factor + add_offset
     """
 
+    @staticmethod
+    def _choose_float_dtype(data, has_offset):
```
Comment (shoyer) [inline xarray/coding/variables.py]: Can you make this helper function use `dtype` as an argument instead of `data`? That would make it clearer that the result dtype is not data dependent.

File: xarray/coding/variables.py [hunk 6]
```diff
@@ -212,11 +212,31 @@ class CFScaleOffsetCoder(VariableCoder):
         decode_values = encoded_values * scale_factor + add_offset
     """
 
+    @staticmethod
```
Comment (shoyer) [inline xarray/coding/variables.py]: This is a small nit, but I prefer not to use `staticmethod` when a normal function will do. We actually say "Avoid `@staticmethod`" in our internal Google style guide (which has gotten out of sync with the public one).

File: xarray/tests/test_coding.py (hunk 4 above)
Comment (Zac-HD) [reply xarray/tests/test_coding.py]: Done.

File: doc/whats-new.rst (hunk 3 above)
Comment (Zac-HD) [reply doc/whats-new.rst]: Done.

**Comment by jhamman** [comment]
This all looks good now. I merged in master and resolved a small conflict in the docs. I'll merge after the tests pass.
